src/Neo/Ledger/SmartThrottler.cs

```diff
@@ -0,0 +1,185 @@
+            _lastResetTime = now;
+            AdjustThrottling(null);
+            _averageFee = CalculateAverageFee(null);
+        }
+
+        // Check if we've hit the tx limit and it's not high priority
+        if ((_transactionsThisSecond >= _maxTransactionsPerSecond) && !IsHighPriorityTransaction(tx))
+            return false;
+
+        _transactionsThisSecond++;
+        return true;
+    }
+
+    /// <summary>
+    /// Updates the network state after a new block is added
+    /// </summary>
+    /// <param name="block">The newly added block</param>
+    public void UpdateNetworkState(Block block)
+    {
+        var currentTime = TimeProvider.Current.UtcNow.ToTimestampMS();
+        var blockTime = currentTime - _lastBlockTimestamp;
+
+        _recentBlockTimes.Enqueue(blockTime);
+        if (_recentBlockTimes.Count > BlockTimeWindowSize)
+            _recentBlockTimes.Dequeue();
+
+        _lastBlockTimestamp = currentTime;
+        _unconfirmedTxCount = _memoryPool.Count;
+        _averageFee = CalculateAverageFee(block);
+
+        AdjustThrottling(block);
+    }
+
+    /// <summary>
+    /// Adjusts throttling parameters based on current network conditions
+    /// </summary>
+    private void AdjustThrottling(Block? block)
+    {
+        var memoryPoolUtilization = (double)_memoryPool.Count / _system.Settings.MemoryPoolMaxTransactions;
+        var networkLoad = EstimateNetworkLoad(block);
+
+        _maxTransactionsPerSecond = CalculateOptimalTps(memoryPoolUtilization, networkLoad, block);
+    }
+
+    /// <summary>
+    /// Estimates current network load
+    /// </summary>
+    /// <returns>An integer between 0 and 100 representing the estimated network load</returns>
+    private int EstimateNetworkLoad(Block? block)
+    {
+        var load = 0;
+
+        // 1. Memory pool utilization (30% weight)
+        var memPoolUtilization = (double)_memoryPool.Count / _system.Settings.MemoryPoolMaxTransactions;
+        load += (int)(memPoolUtilization * 30);
+
+        // 2. Recent block times (30% weight)
+        if (_recentBlockTimes.Count > 0)
+        {
+            var avgBlockTime = _recentBlockTimes.Average(t => (double)t);
+            var blockTimeRatio = avgBlockTime / _system.Settings.MillisecondsPerBlock;
+            load += (int)(Math.Min(blockTimeRatio, 2) * 30); // Cap at 60 points
+        }
+
+        // 3. Current block transaction count or unconfirmed transaction growth rate (40% weight)
+        if (block != null)
+        {
+            var blockTxRatio = (double)block.Transactions.Length / _system.Settings.MaxTransactionsPerBlock;
+            load += (int)(Math.Min(blockTxRatio, 1) * 40);
+        }
+        else
+        {
+            var txGrowthRate = (double)_unconfirmedTxCount / _system.Settings.MaxTransactionsPerBlock;
+            load += (int)(Math.Min(txGrowthRate, 1) * 40);
+        }
+
+        return Math.Min(load, 100); // Ensure load doesn't exceed 100
```

Comment:
This is bad for any weight metric, insert some cut after waiting.

This part of the code needs enhanced.

If we decide move on with this solution I can later test and analyze.

Reply:
Updated, now maximum load is 100.

Reply:
> how could it be more than 100?

The same reason you have this `Math.Min(load, 100);` here